### PR TITLE
[[ Perf ]] Optimization of native string processing

### PR DESCRIPTION
**Pull request by runrevmark**
This patch optimizes string processing for native strings in a number of ways.

It contains a re-implemented stack of low-level native string operations which take full advantage of the properties of the native encoding.

It adds a libfoundation function for 'delimited' chunk offset operations and specializations for native strings with 1 char delimiters (the most common case). These are now used by lineOffset and itemOffset.

It adds a libfoundation function for 'delimited' chunk range operations. These are now used for evaluation of 'item x [ to y ] of' and 'line x [ to y ] of'.

[Review on Reviewable](https://reviewable.io/reviews/livecode/livecode/3267)

### CODE REVIEW

**Comment by peter-b** [comment]
Well this looks really really nice. :grinning: I'm really looking forward to seeing the finished version!

**Comment by peter-b** [comment]
Possible issues: this patch will currently only work with GCC >= 4.7 and C++11, because in earlier versions of C++ function arguments to templates are required to have external linkage, i.e. non-`static`.

**Comment by livecodeali** [comment]
This is a reminder to find and replace occurrences of 'occurrance' with 'occurrence' :-)

File: engine/src/chunk.cpp [hunk 1]
```diff
@@ -2283,6 +2283,191 @@ static void skip_word(const char *&sptr, const char *&eptr)
...
+        return;
+    
+    if (codeunit != nil &&
+        !__MCCRefMarkChunkForEval(CT_CODEUNIT, ctxt, codeunit, x_mark))
+        return;
+        
+    if (byte != nil &&
+        !__MCCRefMarkChunkForEval(CT_BYTE, ctxt, byte, x_mark))
+        return;
+}
```
Comment (peter-b) [inline engine/src/chunk.cpp:2469]: Please add an fallback (e.g. `MCUnreachable()`) for if execution falls through all these tests.

Comment (runrevmark) [inline engine/src/chunk.cpp:2469]: There is no fallback here... It just processes each potential chunk type in turn - if there are none, the range doesn't get restricted.

**Comment by peter-b** [comment]
Reviewed 5 of 12 files at r1, 1 of 7 files at r2, 1 of 5 files at r3, 4 of 11 files at r4.
Review status: 11 of 18 files reviewed at latest revision, 5 unresolved discussions.

---

_[libfoundation/include/foundation.h, line 1929 [r4]](https://reviewable.io:443/reviews/livecode/livecode/3267#-K3P-FJFNwWhlmS3IgMO:-K3P-FJFNwWhlmS3IgMP:628271416) ([raw file](https://github.com/livecode/livecode/blob/402409e08f5c395a93a8f98880b3837e92f43c03/libfoundation/include/foundation.h#L1929)):_
If `2` vs `2U` doesn't make a difference, maybe revert these changes.

---

_[libfoundation/src/foundation-string-native.cpp.h, line 91 [r4]](https://reviewable.io:443/reviews/livecode/livecode/3267#-K3P5jwPvbPvWDXWMSYg:-K3P5jwPvbPvWDXWMSYh:-1830891449) ([raw file](https://github.com/livecode/livecode/blob/402409e08f5c395a93a8f98880b3837e92f43c03/libfoundation/src/foundation-string-native.cpp.h#L91)):_
Maybe `kMCStringFoldWindows1252_Mapping` and `kMCStringFoldMacRoman_Mapping` should be explicitly `const`.

---

_[libfoundation/src/foundation-string-native.cpp.h, line 101 [r4]](https://reviewable.io:443/reviews/livecode/livecode/3267#-K3P8DsX83KUITpbnOmM:-K3P8DsX83KUITpbnOmN:1045573882) ([raw file](https://github.com/livecode/livecode/blob/402409e08f5c395a93a8f98880b3837e92f43c03/libfoundation/src/foundation-string-native.cpp.h#L101)):_
I don't think this is correct.  For example, `{` (`0x7B`) is not cased, so surely `__MCNativeChar_CheckedFold()` should return false for it.

---

_[libfoundation/src/foundation-string-native.cpp.h, line 1065 [r4]](https://reviewable.io:443/reviews/livecode/livecode/3267#-K3PCBdrIY3co5gvocvd:-K3PCBdrIY3co5gvocve:-113685301) ([raw file](https://github.com/livecode/livecode/blob/402409e08f5c395a93a8f98880b3837e92f43c03/libfoundation/src/foundation-string-native.cpp.h#L1065)):_
The argument should be `char_t *x_chars` because it is mutated.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/3267)_

**Comment by peter-b** [comment]
Review status: 11 of 18 files reviewed at latest revision, 4 unresolved discussions.

---

_[libfoundation/include/foundation.h, line 1929 [r4]](https://reviewable.io:443/reviews/livecode/livecode/3267#-K3P-FJFNwWhlmS3IgMO:-K3PDxSQ6XfX4w6qzoAl:-1133596915) ([raw file](https://github.com/livecode/livecode/blob/402409e08f5c395a93a8f98880b3837e92f43c03/libfoundation/include/foundation.h#L1929)):_
(Or modify them; obviously the bit constants are still needed!)

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/3267)_

**Comment by peter-b** [comment]
Reviewed 1 of 11 files at r4, 7 of 7 files at r5.
Review status: 17 of 18 files reviewed at latest revision, 3 unresolved discussions.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/3267)_

**Comment by runrevmark** [comment]
Review status: 17 of 18 files reviewed at latest revision, 1 unresolved discussion.

---

_[libfoundation/src/foundation-string-native.cpp.h, line 101 [r4]](https://reviewable.io:443/reviews/livecode/livecode/3267#-K3P8DsX83KUITpbnOmM:-K3PFEDwHKy2ckE46ccV:1083930167) ([raw file](https://github.com/livecode/livecode/blob/402409e08f5c395a93a8f98880b3837e92f43c03/libfoundation/src/foundation-string-native.cpp.h#L101)):_
Well, it is a performance optimization if CheckedFold returns false - we need to create the appropriate table at some point to use. I'll add a TODO comment about it.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/3267)_

**Comment by peter-b** [comment]
This is going to be :sparkles:awesome:sparkles:

@livecode-vulcan review ok 9b18a1c

-WIP

---

Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/3267#-:-K3PVChV3CQMnemDXxo4:1414285563)_

**Comment by livecode-vulcan** [comment]
💙 review by @peter-b ok 9b18a1cc472fa7a11b091dc8fea94e7263c3a44f

**Comment by peter-b** [comment]
@livecode-vulcan review ok 1962396

---

Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/3267#-:-K3PZnJW7ZHRT0rxx6pd:-1167224084)_

**Comment by livecode-vulcan** [comment]
💙 review by @peter-b ok 19623961fd0e2f14bef0cd2d1e7fccd5b6a8ff4b

**Comment by livecode-vulcan** [comment]
😎 test success 19623961fd0e2f14bef0cd2d1e7fccd5b6a8ff4b
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1195)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1206)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1192)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1199)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1210)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1212)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1214)

**Pull request merged**